C#: MSBuild logs and panel enhancements

paulloz opened the pull request:
I'm continuing my quest about improving the C# experience.

- Rename the current .NET related editor settings from `mono` to `dotnet`, as it become more and more confusing for people to see the word mono in the interface.
  ![image](https://user-images.githubusercontent.com/437025/214638010-27d47872-9940-4882-b010-8c2afee0b949.png)
- Add a button on the right of the MSBuild panel to directly open the logs folder, which is currently quite hard to pinpoint (since it uses an MD5 hash in its path).
  ![image](https://user-images.githubusercontent.com/437025/214638446-90e50f3f-8e29-40b8-91b1-c218f8158139.png)
- Add a few settings, so the user is able to tune how MSBuild logs behave.
  ![image](https://user-images.githubusercontent.com/437025/214638873-5cf92ac1-42e1-4755-83d2-a7603b5281e0.png)
  - **Verbosity Level** is tied to the `-v` flag.
  - **No Console Logging** switches `-noconlog` on or off, basically either print to the console and the logfile or to the logfile alone.
  - **Create Binary Log** switches `-bl` on or off, binary logs being a very useful tool to debug MSBuild issues.
- Finally, as I suggested in #71326, try to tie the language used by MSBuild with the language used by Godot. Won't solve that issue entirely, but is still a step in the right direction IMHO.
  |en|fr|ja|
  |--|--|--|
  |![image](https://user-images.githubusercontent.com/437025/214640344-a1dac804-9ea7-483b-ba4c-37c1d2766947.png)|![image](https://user-images.githubusercontent.com/437025/214640539-8c4e6d12-3a11-4a8b-b5bf-4da6beb92433.png)|![image](https://user-images.githubusercontent.com/437025/214640912-3ae47190-30b5-4a53-9afd-c57db3c82999.png)|

paulloz committed:
Rename existing editor settings `mono` -> `dotnet`

paulloz committed:
Add button to open the msbuild logs folder

paulloz committed:
Try and match MSBuild and Godot UI languages

raulsntos commented on modules/mono/editor/GodotTools/GodotTools/GodotSharpEditor.cs:
```diff
@@ -171,7 +179,7 @@ public void ShowErrorDialog(string message, string title = "Error")
         public Error OpenInExternalEditor(Script script, int line, int col)
         {
-            var editorId = (ExternalEditorId)(int)_editorSettings.GetSetting("mono/editor/external_editor");
+            var editorId = (ExternalEditorId)(int)_editorSettings.GetSetting(Settings.ExternalEditor);
```
Users coming from previous betas will lose their settings.

paulloz replied:
How do you think this should be handled?

raulsntos replied:
I don't know if we have something in Godot to handle editor setting renames. If we don't, we may want to still look for the old name and, if found, set the new setting's value; at least for the betas/RCs.

cc @akien-mga

neikeq replied:
Unless there's a built-in way to do this, I think it may be too much work to bother with it. Otherwise, you could check if the old setting is present, move its value to the new one, and remove the old one to avoid repeating this.

paulloz replied:
Can you be more specific about what you mean by "not bother with it"? Are you implying we'd leave everything labelled "mono" or that we could afford to break settings compatibility between betas?

neikeq replied:
Sorry, I meant not bother about migrating from the old setting. It may be annoying, but it's a one-time thing and there are other breaking changes coming as well.

paulloz commented on modules/mono/editor/GodotTools/GodotTools/Build/BuildSystem.cs:
```diff
@@ -222,6 +249,16 @@ private static void AddLoggerArgument(BuildInfo buildInfo, Collection<string> ar
+                return;
+
+            arguments.Add($"-bl:{Path.Combine(buildInfo.LogsDirPath, "msbuild.binlog")}");
+            arguments.Add("-ds:False"); // Honestly never understood why -bl also switches -ds on.
```
I guess not, it just spams the small area we have for logs in the editor. And the info is anyway embedded in the binlog.

raulsntos replied:
I guess if the info is embedded anyway and this only affects the MSBuildPanel it makes sense to hide it.

paulloz replied:
Basically: every other format is redundant when you produced a binlog, since you can use it to reconstruct any other log after the fact.

paulloz commented on modules/mono/editor/GodotTools/GodotTools/GodotSharpEditor.cs:
```diff
@@ -510,6 +532,24 @@ public override void _EnablePlugin()
             AddChild(GodotIdeManager);
         }
 
+        public override void _DisablePlugin()
+        {
+            base._DisablePlugin();
+
+            _editorSettings.SettingsChanged -= OnSettingsChanged;
+        }
+
+        private void OnSettingsChanged()
+        {
+            // We want to force NoConsoleLogging to true when the VerbosityLevel is at Detailed or above.
+            // At that point, there's so much info logged that it doesn't make sense to display it in
+            // the tiny editor window, and it'd make the editor hang or crash anyway.
+            var verbosityLevel = (VerbosityLevelId)(int)_editorSettings.GetSetting(Settings.VerbosityLevel);
+            var hideConsoleLog = (bool)_editorSettings.GetSetting(Settings.NoConsoleLogging);
+            if (verbosityLevel >= VerbosityLevelId.Detailed && !hideConsoleLog)
+                _editorSettings.SetSetting(Settings.NoConsoleLogging, Variant.From(true));
+        }
```
You mean forcing `-noconlog` regardless of `Settings.NoConsoleLogging` if the verbosity is too high? I thought about it, but it felt better to show the user the checkbox getting automatically set to true. Makes it more "intentional"?

raulsntos replied:
Yes, that's what I meant. I'm not sure what's better to be honest.

paulloz commented on modules/mono/editor/GodotTools/GodotTools/Build/BuildSystem.cs:
```diff
@@ -35,6 +38,8 @@ public static class BuildSystem
             startInfo.UseShellExecute = false;
             startInfo.CreateNoWindow = true;
+            startInfo.EnvironmentVariables["DOTNET_CLI_UI_LANGUAGE"]
+                = (string)editorSettings.GetSetting("interface/editor/editor_language");
```
Now yes: [example based on all the po files in the repo](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyA6AcQBsJgBDdgSwBe3DHwh4A3NWoA3blHpgc7DDlgBnegF56eGAHcA2gF16Ab3oAibgDMLaS3LsOBTi8ADmr4Hi9RXYbn81O2p6MPCIyKiwiwATQPs4mFcYdhS8AH0AZQALbj0UiBTgxJg1DIBBACUQqmj6+osYDBScV2sEy2s+dr40xOs/RPdOi3d+0Iap8Isc5MScnoWhyxy2xL5Y1z5PDZLLPha0SempiwArUYBra4BbVyuixOVXdmlXW6uPibrTs9uVhZbvsLHhgK48P1LAAHKEWaGA6FHGEYDIAIRqx1+f0aUCellwrjUS0sai+iTUcLUAEciVAMgBhACeUCp9IAMsIfFicQ0LGp3okMPNLBgcq4MHCMICcOTLKoMgAFADSrmkJIsAhyjIAcrVeY0tRkABKqxJGgAqAHUnABfSRUajWaAwbhgHL0AAUsnkimUqhg9D4eAUShU6gAlNRTCdGGQAJye3R6egMsMBgCSeGdnr94ZgzCqMFhbpgnoA5Bly/Zy3ByxGI8wAKJ4cZ8NQ5HXcW4wCMO21AA===). Worst case is for `pr` where the locale is found but is still weirdly named "pr"? The CLI will automatically revert to English if by any chance there's an issue (e.g. I believe if the language pack is not installed on your machine).

paulloz commented on modules/mono/editor/GodotTools/GodotTools/Build/BuildInfo.cs:
```diff
@@ -22,8 +22,7 @@ public sealed partial class BuildInfo : RefCounted // TODO Remove RefCounted onc
 
-        public string LogsDirPath =>
-            Path.Combine(GodotSharpDirs.BuildLogsDirs, $"{Solution.MD5Text()}_{Configuration}");
+        public string LogsDirPath => GodotSharpDirs.LogsDirPathFor(Configuration);
```
I need to stop being a 🤡

neikeq commented:
> * **No Console Logging** switches `-noconlog` on or off, basically either print to the console and the logfile or to the logfile alone.

What is the use case for this?

> * **Create Binary Log** switches `-bl` on or off, binary logs being a very useful tool to debug MSBuild issues.

Does this override our custom MSBuild logger? We rely on that one to show the list of errors and warnings after building.

paulloz commented:
> What is the use case for this?

Being able to produce useful build logs (on disk) without flooding the console.

> Does this override our custom MSBuild logger? We rely on that one to show the list of errors and warnings after building.

This doesn't override anything, the custom logger hasn't been touched.

raulsntos reviewed (approved):
LGTM

paulloz committed:
Add new settings about MSBuild

- toggle creation of binary logs
- manage log verbosity
- toggle logging in console

paulloz commented:
Rewrote the commit history as discussed. Shouldn't change anything behaviourwise.

akien-mga commented:
Thanks!

DmitriySalnikov commented:
is it related? #64276

paulloz commented:
@DmitriySalnikov Do you mean "does it fix it?"? If so, not sure. All we're doing here is forcing the CLI language to the same locale as Godot. I didn't see any encoding issue with all the locales I had available to test, but didn't explicitly check ru.
